neoemf-core/src/main/java/fr/inria/atlanmod/neoemf/core/internal/LazyStoreList.java

```diff
@@ -0,0 +1,416 @@
+/*
+ * Copyright (c) 2013-2018 Atlanmod, Inria, LS2N, and IMT Nantes.
+ *
+ * All rights reserved. This program and the accompanying materials are made
+ * available under the terms of the Eclipse Public License v2.0 which accompanies
+ * this distribution, and is available at https://www.eclipse.org/legal/epl-2.0/
+ */
+
+package fr.inria.atlanmod.neoemf.core.internal;
+
+import fr.inria.atlanmod.neoemf.core.PersistentEObject;
+import fr.inria.atlanmod.neoemf.data.store.Storable;
+import fr.inria.atlanmod.neoemf.data.store.Store;
+import fr.inria.atlanmod.neoemf.data.store.adapter.StoreAdapter;
+
+import org.eclipse.emf.common.notify.Notification;
+import org.eclipse.emf.common.util.ECollections;
+import org.eclipse.emf.ecore.EStructuralFeature;
+import org.eclipse.emf.ecore.InternalEObject;
+import org.eclipse.emf.ecore.util.DelegatingEcoreEList;
+import org.eclipse.emf.ecore.util.EcoreEList;
+
+import java.util.Arrays;
+import java.util.Collection;
+import java.util.Iterator;
+import java.util.List;
+import java.util.ListIterator;
+import java.util.stream.Collectors;
+
+import javax.annotation.Nonnull;
+import javax.annotation.ParametersAreNonnullByDefault;
+
+import static fr.inria.atlanmod.commons.Preconditions.checkPositionIndex;
+import static java.util.Objects.isNull;
+
+/**
+ * A {@link List} representing a multi-valued feature which behaves as a proxy and that delegates its operations to the
+ * associated {@link Store}.
+ *
+ * @param <E> the type of elements in this list
+ *
+ * @see PersistentEObject#eStore()
+ */
+@ParametersAreNonnullByDefault
+// TODO Use a local cache (Lazy<List<Object>>) to avoid atomic operations.
+// Only when the resource is NOT distributed, and when the full load is beneficial !
+// - Retrieve all values first (getAll)
+// - Cache them
+// - Execute operations on the cache
+// Think to update the cached values on write operations (only if the local cache is loaded (Lazy#ifLoaded))
+// Think to adapt the immutable list returned by #delegateList()
+public class LazyStoreList<E> extends DelegatingEcoreEList.Dynamic<E> {
```

Comment:
![MAJOR](https://sonarsource.github.io/sonar-github/severity-major.png 'Severity: MAJOR') This class has 10 parents which is greater than 5 authorized. [![rule](https://sonarsource.github.io/sonar-github/rule.png)](https://sonarcloud.io/coding_rules#rule_key=squid%3AMaximumInheritanceDepth)
![MINOR](https://sonarsource.github.io/sonar-github/severity-minor.png 'Severity: MINOR') Override the "equals" method in this class. [![rule](https://sonarsource.github.io/sonar-github/rule.png)](https://sonarcloud.io/coding_rules#rule_key=squid%3AS2160)